tests/s2n_test.h

```diff
@@ -200,7 +200,16 @@ int test_count;
 /* for use with S2N_RESULT */
 #define EXPECT_OK( function_call )  EXPECT_TRUE( s2n_result_is_ok(function_call) )
 
-#define EXPECT_BYTEARRAY_EQUAL( p1, p2, l ) EXPECT_EQUAL( memcmp( (p1), (p2), (l) ), 0 )
+/* Memcmp is annotated with a non-null attribute. The if statement prevents
+ * calling memcmp with NULL arguments when the length is zero.
+ */
```

Comment:
What do you mean by "annotated with a non-null attribute"? 

And wouldn't this work better on the if statement itself, rather than calling out the if statement from several contexts up?

Reply:
According to the `memcmp` definition:
```
/* Compare N bytes of S1 and S2.  */
extern int memcmp (const void *__s1, const void *__s2, size_t __n)
     __THROW __attribute_pure__ __nonnull ((1, 2));
```

There is a `__nonnull` attribute, which is what I mean. UBSAN will complain if this attribute is violated.

The GCC Doc specifies:

> The nonnull attribute specifies that some function parameters should be non-null pointers. 

Source: https://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Function-Attributes.html#:~:text=nonnull%20(arg,len)%0A%20%20%20%20%20%20%20%20%20%20%09__attribute__((nonnull))%3B

I agree with your second point. That will be adopted in the next commit.

Reply:
There isn't a single implementation of memcmp, although there are standards that define behavior. I believe memcmp with NULL is undefined according to the c99 standard, but that's worth checking. I'm definitely sure not all implementations use the GCC nonnull attribute :)